### PR TITLE
Merge typing of expressions and formulas

### PR DESCRIPTION
**Pull request by strub**

### CODE REVIEW

File: src/ecParser.mly [hunk 1]
```diff
@@ -917,139 +896,8 @@ tyvar_annot:
...
-| e1=loc(expr_chained_orderings) op=loc(ordering_op) ti=tvars_app? e2=expr
-    { let (lce1, (e1, le)) = (e1.pl_loc, unloc e1) in
-      let loc = EcLocation.make $startpos $endpos in
-        (peapp_symb loc "&&" None
-           [EcLocation.mk_loc lce1 e1;
-            EcLocation.mk_loc loc
-              (peapp_symb op.pl_loc (unloc op) ti [le; e2])],
-         e2) }
+%inline sexpr: f=sform { mk_loc f.pl_loc (Expr f) }
+%inline  expr: f=form  { mk_loc f.pl_loc (Expr f) }
```
Comment (bgregoir) [inline src/ecParser.mly:900]: This clearly simplify parsing :-)

File: src/ecParsetree.ml [hunk 2]
```diff
@@ -170,7 +122,37 @@ type glob_or_var =
...
+  | `Full of (ppattern * pstmt) list
+  | `If   of (ppattern * pstmt) * pstmt option
+]
+
+and pscond = pexpr * pstmt
+and pinstr = pinstr_r located
+and pstmt  = pinstr list
+
+and pexpr   = pexpr_r located
+and pexpr_r = Expr of pformula
```
Comment (bgregoir) [inline src/ecParsetree.ml:153]: It is really useful to have the Expr constructor ?

File: src/ecTyping.ml [hunk 3]
```diff
@@ -2811,7 +2578,7 @@ and trans_gbinding env ue decl =
   in snd_map List.flatten (List.map_fold trans1 env decl)
 
 (* -------------------------------------------------------------------- *)
-and trans_form_or_pattern env ?mv ?ps ue pf tt =
+and trans_form_or_pattern env mode ?mv ?ps ue pf tt =
```
Comment (bgregoir) [inline src/ecTyping.ml:2581]: I assume mode allows to locate the error (not an expression) more precisely ?

**Review by bgregoir — APPROVED** [review]
This clearly simplify...

File: src/ecParsetree.ml (hunk 2 above)
Comment (strub) [reply src/ecParsetree.ml:153]: Yes. Internally, expressions & formulas deserve different purposes and the high-level tactic API takes parse trees. This forbids using one in place of the other.

File: src/ecTyping.ml (hunk 3 above)
Comment (strub) [reply src/ecTyping.ml:2581]: The notion of `mode` is not new. It has been used as a parameter for the operator resolution mechanism (and so was hard-coded to `Form` in this function). Now that the function is here for translating formulas & expressions, it must be passed as a parameter. And yes, it is also used to inform the function whether an expression or formula is translated and to raise an error when a formula is not a valid expression.

**Pull request merged**